akka-distributed-data/src/main/scala/akka/cluster/ddata/ORMap.scala

```diff
@@ -174,8 +303,123 @@ final class ORMap[A, B <: ReplicatedData] private[akka] (
         case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
       }
     }
+    val removedKeysIterator = (tombstoneVals.keys ++ that.tombstoneVals.keys).iterator
+    var removedValues = Map.empty[A, B]
+    removedKeysIterator.foreach { key ⇒
+      (this.tombstoneVals.get(key), that.tombstoneVals.get(key)) match {
+        case (Some(thisValue), Some(thatValue)) ⇒
+          if (thisValue.getClass != thatValue.getClass) {
+            val errMsg = s"Wrong type for merging [$key] in [${getClass.getName}], existing type " +
+              s"[${thisValue.getClass.getName}], got [${thatValue.getClass.getName}]"
+            throw new IllegalArgumentException(errMsg)
+          }
+          val removedValue = thisValue.merge(thatValue.asInstanceOf[thisValue.T]).asInstanceOf[B]
+          removedValues = removedValues.updated(key, removedValue)
+        case (Some(thisValue), None) ⇒
+          removedValues = removedValues.updated(key, thisValue)
+        case (None, Some(thatValue)) ⇒
+          removedValues = removedValues.updated(key, thatValue)
+        case (None, None) ⇒ throw new IllegalStateException(s"missing value for $key")
+      }
+    }
+    // tombstone the values for removed keys
```

Comment:
How long are such tombstone values kept? Looks like removal is triggered by other actions, so do I understand it correctly that if no more updates are performed the tombstone will live forever? I'm worried about growing full state size when different keys are added and removed.

Are there any assumptions that influence correctness with regards to the order or when the full state is delivered/merged and when the delta is delivered/merged. Note that several deltas, e.g. 17, 18, 19 can be grouped together and the full state corresponding to 18 can be applied first and then the delta group 17,18,19 is merged into that.

BTW, it's looking great, I just want to understand that it is rock solid.
I'll take the time to debug what happens in the tests, to get a better understanding.